omegaconf/basecontainer.py

```diff
@@ -257,6 +262,34 @@ def convert(val: Node) -> Any:
 
         assert False
 
+    def _instantiate_structured_config_impl(self, instance_data: Dict[str, Any]) -> Any:
+        """Instantiate an instance of `self._metadata.object_type`, populated by `instance_data`."""
+        from ._utils import get_structured_config_field_names
+
+        object_type = self._metadata.object_type
+        object_type_field_names = set(get_structured_config_field_names(object_type))
+
+        field_items: Dict[str, Any] = {}
+        nonfield_items: Dict[str, Any] = {}
+        for k, v in instance_data.items():
+            if _is_missing_literal(v):
```

Comment:
I don't think that would work. The values of `self` are subconfig objects that are not yet converted to python primitives (possibly DictConfig or ListConfig, and possibly containing unresolved interpolations). The values of `instance_data` are already resolved and converted to dict or list or to dataclass/attrclass instances.

Reply:
The data passed in is derived from the content of self, which is why I am asking it.
It feels redundant.
You have custom code here that is handling things, it can resolve interpolations and deal with containers.

am I missing anything?

Reply:
> The data passed in is derived from the content of self, which is why I am asking it.

This is true.

> It feels redundant.

Take a look at how `instance_data` is calculated in the calling function (`DictConfig._to_content`)... The calculation is nontrivial, and it depends on several things that are defined in the scope of the calling function.

Reply:
Since the calling function has to perform this calculation anyway (for the case where `structured_config_mode == SCMode.DICT`), I think it makes sense to reuse the calculation by passing in `instance_data` for the `structured_config_mode == SCMode.INSTANTIATE` case.

Reply:
I see a few things there that does not make sense when converting a DictConfig backed by a dataclass to an object:
1. Support for not resolving interpolations
2. Support for converting enums to string (doesn't make sense if we are creating an object).

What do you think about this high level approach:
```
# converts THIS dict config to an object if it's backed by a Structured Config (error otherwise):
DictConfig._to_object()
```

The implementation will use `_to_content(scmode=Instantiate)` for any nested containers (ListConfig or DictConfig).
_to_content() can call `DictConfig._to_object()` if the container it is called on is a DictConfig backed by a Structured Config.

Reply:
Hmm... Then what would be the behavior if a client calls like this?
```python
OmegaConf.to_container(cfg, resolve=False, enum_to_str=True, structured_config_mode=SCMode.INSTANTIATE)
```
Such a call is possible with current `to_container` API.

Reply:
I it will not apply resolve and enum_to_str to the instantiated objects (but will apply it to unstructured config nodes).

Reply:
Sure, makes sense to me.
So:
- the structured config gets processed with `enum_to_str=False` and `resolve=False`,
- nested configs inside the structured config get processed with `enum_to_str=False` and `resolve=False`, and
- non-structured parent configs have `enum_to_str` and `resolve` according to keyword args chosen by the client.

Reply:
> Sure, makes sense to me.
> So:
> 
> * the structured config gets processed with `enum_to_str=False` and `resolve=False`,
you mean resolve=True, right?
> * nested configs inside the structured config get processed with `enum_to_str=False` and `resolve=False`, and
> * non-structured parent configs have `enum_to_str` and `resolve` according to keyword args chosen by the client.

Since a Structured Config may contain an unstructured DictConfig, we could consider inheriting the resolve flag for them from the surrounding call, but I am not sure it's worth the added complexity.

in that case, the signature of to_object() will look something like:
```
def to_object(self, resolve_nested_configs: bool, enum_to_str_nested_configs: bool) -> Any:
  ...
```

An alternative is to just say that everything under a Structured Config is always resolved (and enums are not converted to strings).

I think I prefer the second option.

Reply:
>   you mean resolve=True, right?

Yes.

> I think I prefer the second option.

Me too. I'll get started on the diff.

Reply:
Done in https://github.com/omry/omegaconf/pull/502/commits/c672c100f9bfd07f1e753d448118603abd71eea2.